core/shared/src/main/scala/org/typelevel/log4cats/SelfAwareStructuredLogger.scala

```diff
@@ -94,6 +96,90 @@ object SelfAwareStructuredLogger {
       sl.trace(modify(ctx), t)(message)
   }
 
+  def withContextFromKleisli[F[_]: Monad](
+      sl: SelfAwareStructuredLogger[Kleisli[F, Map[String, String], *]]
+  ): SelfAwareStructuredLogger[Kleisli[F, Map[String, String], *]] =
+    withContextF(sl)(
+      Kleisli.ask[F, Map[String, String]]
+    )
+  def withContextF[F[_]: FlatMap](
+      sl: SelfAwareStructuredLogger[F]
+  )(ctx: F[Map[String, String]]): SelfAwareStructuredLogger[F] =
+    new ModifiedContextFSelfAwareStructuredLogger[F](sl)(existingCtx => ctx.map(_ ++ existingCtx))
```

Comment:
I think what we're missing is something like this:

```scala
def fa: F[A]

def faWithContext: F[A] = StructuredLogger[F].locally(Map("k" -> "v"))(fa)
```

so that you can wrap an _effect_, not just a logger. That would help propagate the context to whatever services you're calling, without a need to pass the context or the updated logger.

Reply:
How are you imagining something like `StructuredLogger[F].locally` would work?
Just specific implementations for where there's a `IOLocal` or `Kleisli` around?

Reply:
yeah, I think we'll need concrete implementations

Reply:
I have come to believe that almost all uses of `IOLocal` should have that `cats.mtl.Local` shape.  See https://github.com/typelevel/cats-effect/issues/3385.  And I like the `scope` name: that's what `Local` calls it.